_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,440 @@
+
+## Introduction to Mocks
+
+The basic concept of mocking is replacing real objects with doubles, that we are able to advise how to behave. So to
+say, mocks are stuntmen for our test cases. Each stuntman has his speciality at what he is good at, and so do mocks. In
+software engineering we differentiate in **Test Doubles**. Don't get confused with *Mocks* and *Test Doubles*, we will
+cover the difference in detail later in this article.
+
+Let's imagine we have a service, that processes orders from a database. It's very unhandy to set up a whole database
+just to test that service. In that case we create a *mock* that pretends to be the database, but in the eyes from the
+service it looks like a real database. Furthermore, we can exactly advise the mock how it shall behave. That's exactly
+what stuntmen do. They do the hard and cumbersome work, while we can enjoy doing the nice work! Having this tool, we
+are able to test the service but don't actually need a database.
+
+Here [Mockito](https://www.mockito.org) comes into play. Mockito is a very popular library that supports such testing
+scenarios. Spring also uses Mockito and has its own adapters that provide convenience for developers. Don't worry -
+we don't need to have Spring to use Mockito! Mockito also comes with a standalone library that can be referenced within
+our build tool such as Maven and Gradle.
+
+Consider reading the section [Why Mock?](https://reflectoring.io/spring-boot-mock/#why-mock) for additional information
+about Mocking.
+
+## Different Types of Test Doubles
+
+In the world of code, there are lots of words for test doubles and definitions for their duty. I recommend to
+define a common language within the team.
+
+Here is a little summary of the different types for test doubles and how we use them:
+
+| Type  | Description                          |
+| ----- | ------------------------------------------------------------ |
+| Dummy | A dummy is an object that is often used as a parameter for the method under test but without an actual need of it during the test itself. |
+| Fake  | A fake is an object that has an implementation, but which is restricted and simpler as the real implementation. Due to this restriction, the object is not meant to be used in production. |
+| Stub  | A stub is an object that always returns the same value, regardless of which parameters you provide on a stub's methods. |
+| Mock  | A mock is an object whose behaviour is declared before the test is run. (This is exactly what Mockito is made for!) |
+| Spy   | A spy is an object that logs each method call that is performed on it (including argument values). It can be queried to create assertions in order to verify the behaviour of the system under test. (Spies are supported by Mockito!) |
+
+## Mockito in Use
+
+Consider following example:
+
+![Simple UML Diagram](../assets/img/posts/clean-unit-tests-with-mockito/city-uml-diagram.png)
```

Comment:
Looking at this diagram again...can we remove the `BaseService` and `BaseRepository` classes? They're not referenced anywhere in the text (except in the description of the diagram below) and just make this diagram more complex?